File: cli/python/ime_control/watcher.py
from __future__ import annotations

import time
from collections.abc import Callable, Iterator
from dataclasses import dataclass

from .win32_ime import ImeState, get_foreground_ime_state


@dataclass(frozen=True)
class WatchEvent:
    state: ImeState
    machine_state: str


def state_key(state: ImeState) -> tuple[str, int, int]:
    return (state.machine_state, state.conversion_mode, state.keyboard_layout)


def iter_state_changes(
    *,
    interval: float,
    get_state: Callable[[], ImeState] = get_foreground_ime_state,
) -> Iterator[WatchEvent]:
    previous_key: tuple[str, int, int] | None = None
    while True:
        state = get_state()
        current_key = state_key(state)
        if current_key != previous_key:
            yield WatchEvent(state=state, machine_state=state.machine_state)
            previous_key = current_key
        time.sleep(interval)

File: src/watcher.rs
use crate::win32_ime::ImeState;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct WatchEvent {
    pub state: ImeState,
    pub machine_state: crate::win32_ime::MachineState,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct StateChangeDetector {
    previous_key: Option<StateKey>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
struct StateKey {
    machine_state: crate::win32_ime::MachineState,
    conversion_mode: u32,
    keyboard_layout: u64,
}

impl StateChangeDetector {
    pub fn observe(&mut self, state: ImeState) -> Option<WatchEvent> {
        let key = StateKey {
            machine_state: state.machine_state(),
            conversion_mode: state.conversion_mode,
            keyboard_layout: state.keyboard_layout,
        };
        if self.previous_key == Some(key) {
            return None;
        }
        self.previous_key = Some(key);
        Some(WatchEvent {
            state,
            machine_state: key.machine_state,
        })
    }
}
